### PR TITLE
Rubocop fixes

### PR DESCRIPTION
**Pull request by stevenjack**
### Problem

Consistency of the code is fairly poor and it would be nice to have a style guide that the gem adheres to.

### Solution

Working on @johnf branch, fix the the set of standard cops then gradually introduce others such as the ruby 1.9 new hash syntax.

Branch based off original start here: https://github.com/johnf/cfndsl/tree/rubocop

### CODE REVIEW

**Comment by stevenjack** [comment]
@johnf Here's the branch I mentioned, would be worth continuing the conversation in here.

I think I tackled the same problem you mentioned in terms of the code duplication and overwriting issue with the `val` key on the two arrays. There's some pretty odd stuff happening in that class and some of the loops as you said don't even you the values of the loop which is pretty pointless.

I think the best course of action is to get all the cops passing, then start working on the bigger problems that we've had to fix like the `lib/cfndsl/types.rb` class which I've abstracted away so it doesn't have the huge amount of duplication as it did before with the `os` ans `aws` types but it's still huge.

File: lib/cfndsl/cloud_formation_template.rb [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            end
+          end
+        end
+      end
+    end
+  end
+
+  class CloudFormationTemplate < OrchestrationTemplate
+    def self.template_types
+      CfnDsl::AWS::Types::Types_Internal
```
Comment (stevenjack) [inline lib/cfndsl/cloud_formation_template.rb]: These `Internal` prefixes aren't going to stay, I need to work out a better name for these otherwise it's `.. AWS::Types::Types` which doesn't really make much sense.

**Comment by johnf** [comment]
@stevenjack Looks like did end up both working at once :)

I'll rebase my work from this morning against this branch shortly.

Yeah agree. Once there is some cleanup I want to get a better understanding of what some of those bits are doing.

With some of the complexity cops. I'll start moving those into the code itself so we can have sensible numbers in ```.rubocop.yaml``` and only old ugly bits of code are excluded till they are cleaned up.

**Comment by johnf** [comment]
@stevenjack A couple more changes you can merge in my branch

**Comment by johnf** [comment]
@stevenjack Have some more commits in my branch. I think the .rubocop.yml is probably down to where we want it.

I've moved some exceptions inline into the code. They are around 5 or so methods that are very complex and at some stage could be improved on.

Probably at a stage where merge could happen and those ones get worked on as a secondary piece of work.

**Comment by stevenjack** [comment]
@johnf Looking good :+1: !

I guess we should try to keep this PR on the smaller side and take a couple of passes on improving the structure and adding more cops in subsequent PRs. I'll start creating some issues that can be addressed in separate, hopefully fairly isolated PRs.

I also want to re-visit the specs as at the moment we only really have a high level integration test and I'd like to get some unit tests in there to test the functionality in isolation.

**Comment by johnf** [comment]
@stevenjack I agree. I think it might also be a good idea to do an rc or prerelease for the next version.

I worry that maybe we've broken some smaller corner cases or something. So it would be good to get more testing out in the wild.

**Comment by kornypoet** [comment]
Big :+1: here. Consistent syntax, more tests (soon hopefully), and then it will be much easier to collaborate on `cfndsl` while keeping the codebase stable.

**Comment by stevenjack** [comment]
@johnf Agree on the potential for a pre-release as I'm also a little tentative about merging this as I haven't got a huge amount of confidence in the current test suite.

Another thought I had was to use [cfndsl-examples](https://github.com/neillturner/cfndsl_examples) and run each template through this branches version of `cfndsl` and diff it against the version that's in the examples..

Not sure about that though, would rather the pre-release - not really sure the best way to promote it though and get people to start using it?

**Comment by stevenjack** [comment]
@kornypoet @johnf I've released the gem as:

`0.5.0.pre`

which can be installed with:

`$:  gem install cfndsl --pre`

What's the best way to get people testing this out?

**Comment by stevenjack** [comment]
I've created an issue for tracking it: https://github.com/stevenjack/cfndsl/issues/171

**Comment by gergnz** [comment]
ping @thebeefcake; we should probably use this as minimum and give it a good run

**Comment by stevenjack** [comment]
Latest version released:

`1.5.1-pre`
